Feature/filter exercises

ungaralex opened the pull request:
<!-- Thanks for contributing to ArTEMiS! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [X] I documented my source code using the JavaDoc / JSDoc style.
- [x] ~I added integration test cases for the server (Spring) related to the features~
- [x] ~I added integration test cases for the client (Jest) related to the features~
- [X] I added screenshots/screencast of my UI changes
- [X] I translated all the newly inserted strings


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
We currently have no way of filtering exercises in the overview (see #386). There should at least be some options of filtering the displayed exercises.

### Description
<!-- Describe your changes in detail -->
I added a new filter dropdown next to the sorting dropdown. The students can now select between multiple options of filtering the displayed exercise list:

- **Select all**: Selects all options
- **Select none**: Selects no options
- **Full score**: Hides/displays exercises for which the students already have an assessment of 100%
- **Overdue**: Hides/displays exercises for which the due date lies in the past

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to ArTEMiS
2. Navigate to any course
3. Click on the filter dropdown next to the sorting dropdown
4. Select any combination of filters and sorting options
5. The view should change accordingly and the exercises should be displayed with the applied filters and in the selected order

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![filter_exercises](https://user-images.githubusercontent.com/27979674/64171233-6f369900-ce52-11e9-9705-3ccf047018cf.gif)

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +68,71 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+     * @param exercise The exercise which should get checked
+     */
+    private needsWork(exercise: Exercise): boolean {
+        console.log(this.user);
```
Please remove console.logs, we don't have it in the linter because we sometimes do it (we could talk about adding it as a rule in my opinion).

ungaralex replied:
removed

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -32,11 +36,15 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
     ngOnInit() {
         this.exerciseCountMap = new Map<string, number>();
+        this.filters = new Set<number>();
+        this.accountService.identity(true).then(user => (this.user = user));
```
I'm not sure if we need the call to the accountService (see next comment), but it would be better to coordinate the asynchronous calls here.
E.g. applyFiltersAndOrder could be called without the user being loaded.

ungaralex replied:
True. I removed the call to the service

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +68,71 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+        console.log(exercise);
+        return (
+            exercise.participations.some(
+                participation => participation.student.id === this.user!.id && participation.results && participation.results.some(result => result.score !== 100),
```
This seems like something to me that the server needs to handle.
Otherwise the student would e.g. be able to see participations in the network tab that are not his.

ungaralex replied:
Yes, the server only loads the relevant exercises, so removed the filter.

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +68,71 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+    }
+
+    /**
+     * Checks if the given exercise still needs work, i.e. is not graded with 100%, or wasn't even started, yet.
```
I wonder if the due date is another important factor here :thinking:

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -15,15 +16,18 @@ import { Exercise, ExerciseService } from 'app/entities/exercise';
     private courseId: number;
     private paramSubscription: Subscription;
     private translateSubscription: Subscription;
+    private filters: Set<number>;
```
Nitpick: I think it would be better called `activeFilters`.

ungaralex replied:
Renamed the variable

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +68,71 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+     * @param filters The filters which should be applied
+     * @param active Should the filters be active or inactive
+     */
+    filterUpdate(filters: number[], active: boolean) {
```
Opinionated:
I think it would be better to not use an active parameter, but to toggle the filters from the activeFilters set.
I would also propose to bind the call `activeFilters.includes(FILTER)` to the input in the template:
`[checked]="activeFilters.includes(FILTER)`.
This way you can be sure that the component and the template are always in sync.

ungaralex replied:
Checked is now bound to the template

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -15,15 +16,18 @@ import { Exercise, ExerciseService } from 'app/entities/exercise';
 export class CourseExercisesComponent implements OnInit, OnDestroy {
     public readonly DUE_DATE_ASC = 1;
     public readonly DUE_DATE_DESC = -1;
+    public readonly OVERDUE = 0;
```
I think it would be better to understand as an enum.
```enum FilterType {
OVERDUE = "OVERDUE", NEEDS_WORK = "NEEDS_WORK"
}```

You can assign enums as a property to components if you want to use them in the template.

If you don't assign strings, the enum members would have numeric values.

ungaralex replied:
Refactored order and filter to enums

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +68,71 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+     * Return the total number of exercises for the logged in user
+     */
+    getNrOfExercises(): number {
+        return [...this.exerciseCountMap.values()].reduce((a, b) => a + b, 0);
```
This should be equivalent to `this.exerciseCountMap.values().length`.

ungaralex replied:
Added sum from lodash

krusche commented:
Today we discussed some changes:
* Show the filters directly with checkboxes: `Hide overdue` and `Hide full score`
* Show the sorting directly with arrows (as done in the table headers).

Both settings should be saved in the local storage of the browser

ungaralex commented:
I created an identical branch in this repository, so we can test the changes on the server. 
We now have two checkboxes (Hide overdue and hide full score). I also removed the sorting dropdown as discussed. I am using a toggle button now, which should fit quite nicely

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.html:
```diff
@@ -70,7 +69,7 @@ <h3 class="text-primary">
             <div class="row mb-1">
                 <div class="col-8">{{ 'artemisApp.courseOverview.exerciseList.details.totalExercises' | translate }}</div>
-                <div class="col-4">{{ course?.exercises.length }}</div>
+                <div class="col-4">{{ getNrOfExercises() }}</div>
```
This can hurt performance as the function is called all the time. It would be better to assign the totalNumber of exercises to a component attribute.

ungaralex replied:
I created a variable

jpbernius commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -2,41 +2,64 @@ import { Component, OnDestroy, OnInit } from '@angular/core';
 
     ngOnInit() {
         this.exerciseCountMap = new Map<string, number>();
+        const filters = localStorage.getItem(SortFilterStorageKey.FILTER);
```
I think we should move interactions with the `localStorage` API into it's  own service. (We could do thin in a follow up)

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -2,41 +2,64 @@ import { Component, OnDestroy, OnInit } from '@angular/core';
     ngOnInit() {
         this.exerciseCountMap = new Map<string, number>();
+        const filters = localStorage.getItem(SortFilterStorageKey.FILTER);
+        this.activeFilters = new Set(filters ? filters.split(',').map(filter => ExerciseFilter[filter]) : []);
```
Hard to read, nested function invocations.
Also ExerciseFilter[filter] could return undefined, which needs to be accounted for.

I would suggest:
```
const filters = localStorage.getItem(SortFilterStorageKey.FILTER) || null;
const filtersInStorage = filters ? filters.split(',').map(filter => ExerciseFilter[filter]).filter(Boolean) : [];
this.activeFilters = new Set(filters); // Make sure filters are unique.
```

ungaralex replied:
Refactored

jpbernius commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -48,10 +71,11 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
 
-        this.translateSubscription = this.translateService.onLangChange.subscribe((event: LangChangeEvent) => {
-            this.groupExercises(this.DUE_DATE_DESC);
+        this.applyFiltersAndOrder();
```
What happens if open Artemis in two windows and change the order in the other window.
This would update the localStorage in an inconsistent matter, wouldn't it?
Maybe, in a new service handling this interaction  (see above) we should [handle the browser events fired for changes in local storage?](https://stackoverflow.com/a/57227771/6324877)

ungaralex replied:
Well, If one window would have checked "Hide overdue" and the other would uncheck it, then it just would be unchecked on a reload of the page. In the end, the most recent change is persisted, so from a user's perspective, I think it still makes sense. But you are definitely right regarding refactoring the local storage into its own service in a future PR

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -2,41 +2,64 @@ import { Component, OnDestroy, OnInit } from '@angular/core';
+        const filters = localStorage.getItem(SortFilterStorageKey.FILTER);
+        this.activeFilters = new Set(filters ? filters.split(',').map(filter => ExerciseFilter[filter]) : []);
+        const order = localStorage.getItem(SortFilterStorageKey.ORDER);
+        this.sortingOrder = !order || order === ExerciseSortingOrder.DUE_DATE_ASC.toString() ? ExerciseSortingOrder.DUE_DATE_ASC : ExerciseSortingOrder.DUE_DATE_DESC;
```
You could make the default clearer:

```
const orderInStorage = localStorage.getItem(SortFilterStorageKey.ORDER);
this.sortingOrder = Object.keys(ExerciseSortingOrder).find(exerciseOrder => exerciseOrder === orderInStorage) || ExerciseSortingOrder.DUE_DATE_ASC;
```

jpbernius commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +84,68 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
 
-    public groupExercises(selectedOrder: number): void {
+    get filterType() {
+        return ExerciseFilter;
```
should this be moved as a `readonly` property to the same place as `ExerciseSortingOrder`?

ungaralex replied:
True, I will refactor it

jpbernius commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +84,68 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+     * Return the total number of exercises for the logged in user
+     */
+    getNrOfExercises(): number {
+        return sum([...this.exerciseCountMap.values()]);
```
Maybe use `Array.from(this.exerciseCountMap.values())`? But TBH I am not sure if this works.

ungaralex replied:
It works, I changed it

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +84,68 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+     * Return the total number of exercises for the logged in user
+     */
+    getNrOfExercises(): number {
+        return sum([...this.exerciseCountMap.values()]);
```
Not part of the scope of this PR as introduced in another PR, but the use of Sets/Maps really makes the control flow a lot harder...

If `this.exerCountMap = {}` this would be: `Object.values(this.exerciseCountMap).reduce(sum, 0)`;

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +84,68 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+     * @param active Should the activeFilters be active or inactive
+     */
+    filterUpdate(filters: ExerciseFilter[]) {
+        filters.forEach(filter => (this.activeFilters.has(filter) ? this.activeFilters.delete(filter) : this.activeFilters.add(filter)));
```
Because of the use of a mutable Map, it would be nicer to use a for loop here, just to make it obvious that the Map is being mutated here.

thilo-behnke replied:
Also I would call the method 'toggleFilters'.

ungaralex replied:
Renamed it

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +84,68 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+     */
+    private needsWork(exercise: Exercise): boolean {
+        return (
+            exercise.participations.some(participation => participation.results && participation.results.some(result => result.score !== 100)) ||
```
Please split this up into two variables.

Also: `participation.results.length === 0` can be replaced with `!participation.results.length`.

ungaralex replied:
I created two variables

thilo-behnke commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +84,68 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+     */
+    private applyFiltersAndOrder() {
+        const filtered = this.course!.exercises.filter(
+            exercise =>
```
Also here please use multiple named variables so that the different criteria can be understood.

ungaralex replied:
I now use some named variables

maxr96 commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.html:
```diff
@@ -1,21 +1,20 @@
+                    </label>
+                    <label class="d-flex align-items-center mb-0 mr-2">
+                        <input type="checkbox" [checked]="activeFilters.has(filterType.NEEDS_WORK)" (change)="filterUpdate([filterType.NEEDS_WORK])" />
+                        <span class="ml-2" jhiTranslate="artemisApp.courseOverview.exerciseList.fullScore">Hide overdue</span>
```
I think you missed this one, should be `Hide full score`

ungaralex replied:
True!

maxr96 commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -2,41 +2,64 @@ import { Component, OnDestroy, OnInit } from '@angular/core';
+}
+
+enum ExerciseSortingOrder {
+    DUE_DATE_ASC = 1,
```
I think it is clearer to have to store them as `ASC` and `DESC` strings and not numbers

ungaralex replied:
True, but we use the numbers to calculate sorting order when sorting the exercises, i.e. get a negative or positive value of the comparison depending on the order.

maxr96 commented on src/main/webapp/app/overview/course-exercises/course-exercises.component.ts:
```diff
@@ -60,15 +84,68 @@ export class CourseExercisesComponent implements OnInit, OnDestroy {
+    /**
+     * Return the total number of exercises for the logged in user
+     */
+    getNrOfExercises(): number {
```
Please type the full name of the method: `getNumberOfExercises`. It is more understandable that way.

ungaralex replied:
I now use the full name

krusche commented:
> I created an identical branch in this repository, so we can test the changes on the server.
> We now have two checkboxes (Hide overdue and hide full score). I also removed the sorting dropdown as discussed. I am using a toggle button now, which should fit quite nicely

Can you please update the screenshots in the original PR description to the new implementation? This makes it much easier for reviewers ;-)

ungaralex commented:
> > I created an identical branch in this repository, so we can test the changes on the server.
> > We now have two checkboxes (Hide overdue and hide full score). I also removed the sorting dropdown as discussed. I am using a toggle button now, which should fit quite nicely
> 
> Can you please update the screenshots in the original PR description to the new implementation? This makes it much easier for reviewers ;-)

I changed the .gif in the description
